flink-yarn/src/test/java/org/apache/flink/yarn/TestingNMClientAsync.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.yarn;
+
+import org.apache.hadoop.yarn.api.records.Container;
+import org.apache.hadoop.yarn.api.records.ContainerId;
+import org.apache.hadoop.yarn.api.records.ContainerLaunchContext;
+import org.apache.hadoop.yarn.api.records.ContainerStatus;
+import org.apache.hadoop.yarn.api.records.NodeId;
+import org.apache.hadoop.yarn.client.api.async.NMClientAsync;
+
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+
+/**
+ * Testing implementation of {@link NMClientAsync} for tests.
+ */
+public class TestingNMClientAsync extends NMClientAsync {
+	private BiFunction<ContainerId, NodeId, ContainerStatus> getContainerStatusAsyncBiFunction = (containerId, nodeId) -> {
+		throw new UnsupportedOperationException("No getContainerStatusAsync function has been set.");
```

Comment:
I think we should try to have a default function that does not throw any exception if possible.

We can change the return type of this function to `Optional<ContainerStatus>`, return `Optional.empty()` by default, and in `getContainerStatusAsync` we check it and only invoke the `callbackHandler` if the function return value is not empty.

Reply:
I'm not sure about that. This function should only be invoked explicitly. If someone needs to invoke it in the test case, they need to define the behavior either.

Reply:
> This function should only be invoked explicitly.
I don't see how is this necessary. We should not assume how RM uses this method.